### ISSUE TITLE
[Task][HF] 2.3: Gradient checkpointing integration with HF transformers

### ISSUE THREAD

**Issue by lvyufeng** [candle-org/candle#229]
## Parent Issue

Part of #204 — HF Ecosystem Roadmap, **Phase 2 (v0.2.0): Task 2.3**

## Description

Ensure `torch.utils.checkpoint.checkpoint` works for both reentrant and non-reentrant paths that transformers uses. Test with `model.gradient_checkpointing_enable()` on tier-1 models.

## Requirements

- [ ] Reentrant checkpointing path works correctly
- [ ] Non-reentrant checkpointing path works correctly (use_reentrant=False)
- [ ] `model.gradient_checkpointing_enable()` integration with HF models
- [ ] `model.gradient_checkpointing_disable()` works correctly
- [ ] Memory savings are observed (re-computation vs storing activations)
- [ ] Gradients are correct through checkpointed segments

## Goal

Remove `test_.*gradient_checkpointing.*` global xfail from compat test suite.

## Key Files

- `src/candle/utils/checkpoint.py` (existing implementation)
- HF transformers integration points

## Success Criteria

```python
import candle as torch
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("gpt2")
model.gradient_checkpointing_enable()
# Training loop should work with reduced memory
```

## Blocked By

None directly.

## Blocks

- Required for memory-efficient training of large models

**Comment by lvyufeng** [comment]
**Cross-roadmap duplicate:** This task overlaps with [Task][Megatron] #258. HF requires reentrant/non-reentrant checkpoint paths; Megatron adds selective recompute and RNG state management. Work should be coordinated.
